swarm/network/discovery_test.go

```diff
@@ -58,3 +65,137 @@ func TestDiscovery(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+// TestSubpeersMsg tests that the correct set of peers is suggested
+// to another peer with a given depth and kademlia
+func TestSubpeersMsg(t *testing.T) {
+
+	// This is the defined depth
+	testDepth := 2
+
+	// construct ProtocolTester and hive
+	params := NewHiveParams()
+	s, hive, err := newHiveTester(t, params, 1, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// register some addresses in specific bins (must coincide with testDepth)
+	registerBzzAddr(0, hive, true)  // bin 0
+	registerBzzAddr(0, hive, true)  // bin 0
+	registerBzzAddr(1, hive, true)  // bin 1
+	registerBzzAddr(1, hive, true)  // bin 1
+	registerBzzAddr(3, hive, true)  // bin 3
+	registerBzzAddr(4, hive, true)  // bin 4
+	registerBzzAddr(3, hive, false) // add a known but not connected peer
+	registerBzzAddr(1, hive, false) // add a known but not connected peer
+
+	// start the hive
+	hive.Start(s.Server)
+	defer hive.Stop()
+
+	// the pivot node is the only one from the ProtocolTester
+	pivot := s.Nodes[0]
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	// we need to wait until the pivot node is actually connected to our hive
+WAIT_PIVOT:
+	for {
+		select {
+		case <-ctx.Done():
+			t.Fatal("Timed out waiting for the pivot node to connect")
+		case <-time.After(100 * time.Millisecond):
+			if _, ok := hive.peers[pivot.ID()]; ok {
+				break WAIT_PIVOT
+			}
+		}
+	}
+
+	// get BzzAddr of the pivot
+	pivotAddress := hive.peers[pivot.ID()]
+	pivotBzz := pivotAddress.BzzAddr.Over()
+
+	// now we need to identify which peers are expected
+	// iterate the hive's connection and only add peers below testDepth
+	var expectedPeers []*BzzAddr
+	hive.EachConn(hive.BaseAddr(), 256, func(p *Peer, po int) bool {
+		if po < testDepth {
+			// don't add the pivot node itself to expectedPeers;
+			// the pivot node was not added as
+			if !bytes.Equal(p.BzzAddr.Over(), pivotBzz) {
+				expectedPeers = append(expectedPeers, p.BzzAddr)
+			}
+		}
+		return true
+	})
+
+	// the test exchange is as follows:
+	// 1. Trigger a subPeersMsg from pivot to our hive
+	// 2. Hive will respond with peersMsg with the set of expected peers
+	err = s.TestExchanges(p2ptest.Exchange{
+		Label: "incoming subPeersMsg",
+		Triggers: []p2ptest.Trigger{
+			{
+				Code: 1,
+				Msg:  &subPeersMsg{Depth: uint8(testDepth)},
```

Comment:
i really think the peer should advertise a different depth than self,, its confusing and deceptive that u use the same